File: app/android/station/src/main/java/app/tankste/station/model/StationModel.kt
package app.tankste.station.model

import app.tankste.navigation.model.CoordinateModel

data class StationModel(
    val id: String,
    val name: String,
    val company: String,
    val address: Address,
    val prices: Prices?,
    val coordinate: CoordinateModel,
    val isOpen: Boolean,
) {

    val label: String
        get() = name.takeIf(String::isNotBlank) ?: company

    data class Address(
        val street: String,
        val houseNumber: String,
        val postCode: String,
        val city: String,
    )

    data class Prices(
        val e5: Double,
        val e5Range: PriceRange,
        val e10: Double,
        val e10Range: PriceRange,
        val diesel: Double,
        val dieselRange: PriceRange
    ) {

        fun getFirstPrice(): Double? =
            e5.takeIf { p -> p != 0.0 }
            ?: e10.takeIf { p -> p != 0.0 }
            ?: diesel.takeIf { p -> p != 0.0 }

        fun getFirstPriceRange(): PriceRange? =
            e5Range.takeIf { r -> r != PriceRange.UNKNOWN }
                ?: e10Range.takeIf { r -> r != PriceRange.UNKNOWN }
                ?: dieselRange.takeIf { r -> r != PriceRange.UNKNOWN }

        enum class PriceRange {
            UNKNOWN,
            CHEAP,
            NORMAL,
            EXPENSIVE,
        }
    }
}

File: station/lib/model/station_model.dart
import 'package:navigation/coordinate_model.dart';
import 'package:station/model/currency_model.dart';

class StationModel {
  final int id;
  final int originId;
  final String name;
  final String brand;
  final StationAddressModel address;
  final CoordinateModel coordinate;
  final bool isOpen;
  final CurrencyModel currency;

  StationModel(
      {required this.id,
      required this.originId,
      required this.name,
      required this.brand,
      required this.address,
      required this.coordinate,
      required this.isOpen,
      required this.currency});

  @override
  String toString() {
    return 'StationModel{id: $id, originId: $originId, name: $name, brand: $brand, address: $address, coordinate: $coordinate, isOpen: $isOpen, currency: $currency}';
  }
}

class StationAddressModel {
  final String street;
  final String houseNumber;
  final String postCode;
  final String city;
  final String country;

  StationAddressModel(
      {required this.street,
      required this.houseNumber,
      required this.postCode,
      required this.city,
      required this.country});

  @override
  String toString() {
    return 'StationAddressModel{street: $street, houseNumber: $houseNumber, postCode: $postCode, city: $city, country: $country}';
  }
}
